[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism

anshul-si opened the pull request:
**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164031
* #163897



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

anshul-si committed:
[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism

[ghstack-poisoned]

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"

**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"

**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"

**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

weifengpy commented on torch/distributed/pipelining/stage.py:
```diff
@@ -641,6 +642,30 @@ def perform_backward(
                     result = perform_backward(backward_type)()
+
+        # If submod is a Replicate module
+        elif isinstance(self.submod, ReplicateModule):
```
I compared ReplicateModule and FSDPModule line by line, but seems they are almost identical. what's the core diffrerence?

anshul-si replied:
The only difference is we need a way to distinguish modules that have been replicated vs fully-sharded so we know whether to get replicate state or fully shard state

H-Huang commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -366,6 +368,238 @@ def apply_tp(
+                group=pp_group,
+            )
+            ref_partial_models = [ref_pipeline_stage.submod]
+            ref_pipeline_schedule = ScheduleClass(
```
Your loss function is y.sum(), so I think you need `scale_grads=False`

H-Huang commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -366,6 +368,238 @@ def apply_tp(
+                ref_pipeline_schedule.step(target=labels, losses=ref_losses)
+
+                for loss, ref_loss in zip(losses, ref_losses):
+                    self.assertEqual(loss, ref_loss)
```
In a lot of our tests we compare single-device model grads with pipeline parallel model grads. This test just compares the loss of two pipelines. Might be out of scope to add this though, but maybe can add a TODO

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"

**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[Replicate][Pipeline Parallelism] integration of new replicate function with pipeline parallelism"

**Summary:** In order to test numerics for replicate + pp, stage.py needs to be able to call replicate's backward manually as pipeline parallelism doesn't have this feature. 

**Test Case**
1.  pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

weifengpy reviewed (approved):
looking good on the fsdp part

H-Huang reviewed (approved):
Looks good!
